### PR TITLE
fix(browser): run `toMatchScreenshot` only once when used with `expect.element`

### LINKED ISSUE #8959: Infinite loop of screenshot creation on `toMatchScreenshot` failure
### Describe the bug

Vitest starts creating reference screenshots in an infinite loop when my test fails.

Expected vitest to take a single screenshot on `toMatchScreenshot` failure.

### Reproduction
https://github.com/Romej/vitest-browser-infinite-screenshots#

Image

### System Info

```shell
System:
    OS: macOS 15.7
    CPU: (14) arm64 Apple M4 Pro
    Memory: 3.52 GB / 48.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 22.19.0
    npm: 10.9.3
    pnpm: 10.17.1
```

### Used Package Manager

npm

### Validations

- [x] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [x] Read the [docs](https://vitest.dev/guide/).
- [x] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [x] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [x] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by macarie**
### Description

Fixes #8959

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.
- [ ] Please check [Allow edits by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to make review process faster. Note that this option is not available for repositories that are owned by Github organizations.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/vitest/src/integrations/chai/poll.ts [hunk 1]
```diff
@@ -69,12 +69,36 @@ export function createExpectPoll(expect: ExpectStatic): ExpectStatic['poll'] {
...
+                try {
+                  resolve(await assertionFunction.call(assertion, ...args))
+                }
+                catch (err) {
+                  if (chai.util.flag(assertion, '_poll.assert_once')) {
+                    clearTimeout(intervalId)
+                    clearTimeout(timeoutId)
+
+                    rejectWithCause(err)
+                  }
```
Comment (macarie) [inline packages/vitest/src/integrations/chai/poll.ts]: Not entirely convinced by this solution, it adds _yet-another_ exit point to a function that already has many...

**Comment by macarie** [comment]
Pushed a refactor of the poll assertion logic from the nested timer callbacks to a linear async/await loop.

The control flow was quite hard to follow, there were multiple timeouts to track and exit points across different scopes. This also fixes a potential race condition where `check` could run concurrently when the timeout fired while the previous call was mid-`await`.

The behavior is _essentially_ unchanged, the only difference is the last attempt now waits `interval` before running and the assertion succeeds if it is successful (could keep this behavior, but there were edge cases in which `lastError` could still be `undefined`, thus throwing when attempting to set `cause`).

Happy to revert if there are concerns.

File: test/browser/fixtures/expect-dom/toMatchScreenshot.test.ts [hunk 2]
```diff
@@ -421,4 +421,114 @@ describe('.toMatchScreenshot', () => {
...
+          <actual>
+
+        Diff image:
+          <diff>
+      `)
+
+      const elapsed = performance.now() - start
+
+      // Allow some buffer for screenshot capture + diffing
+      expect(elapsed).toBeLessThan(renderDelay + 500)
```
Comment (macarie) [inline test/browser/fixtures/expect-dom/toMatchScreenshot.test.ts]: Locally, 500ms are enough for all browsers to finish. Might be worth monitoring on CI to see if it's flaky.

File: test/browser/fixtures/expect-dom/toMatchScreenshot.test.ts [hunk 3]
```diff
@@ -421,4 +421,114 @@ describe('.toMatchScreenshot', () => {
...
+
+      let errorMessage: string
+
+      try {
+        await expect.element(locator).toMatchScreenshot()
+      } catch (error) {
+        errorMessage = error.message
+      }
+
+      expect(typeof errorMessage === 'string')
```
Comment (sheremet-va) [inline test/browser/fixtures/expect-dom/toMatchScreenshot.test.ts]: This doesn't do anything, did you mean `expect.assert`? We should maybe throw an error if the methods are not called 😄

Comment (macarie) [reply test/browser/fixtures/expect-dom/toMatchScreenshot.test.ts]: Editing fail on my part, it was an `if` before but forgot to change `===` with `).toBe(` 😅

Good catch, thanks!

File: test/browser/fixtures/expect-dom/toMatchScreenshot.test.ts (hunk 2 above)
Comment (macarie) [reply test/browser/fixtures/expect-dom/toMatchScreenshot.test.ts]: Looks like it's failing in CI. Maybe I can just check `< 30_000` since that's poll's default timeout.

**Review by sheremet-va — APPROVED** [review]
